ARCs/arc-0056.md

```diff
@@ -0,0 +1,343 @@
+    create: ("NoOp" | "OptIn" | "DeleteApplication")[];
+    /** OnCompletes this method allows when appID !== 0 */
+    call: (
+      | "NoOp"
+      | "OptIn"
+      | "CloseOut"
+      | "ClearState"
+      | "UpdateApplication"
+      | "DeleteApplication"
+    )[];
+  };
+  /** If this method does not write anything to the ledger (ARC-22) */
+  readonly?: boolean;
+  /** ARC-28 events that MAY be emitted by this method */
+  events?: Array<Event>;
+  /** Information that clients can use when calling the method */
+  recommendations?: {
+    /** The number of inner transactions the caller should cover the fees for */
+    innerTransactionCount?: number;
+    /** Recommended box references to include */
+    boxes?: {
+      /** The app ID for the box */
+      app?: number;
+      /** The base64 encoded box key */
+      key: string;
+      /** The number of bytes being read from the box */
+      readBytes: number;
+      /** The number of bytes being written to the box */
+      writeBytes: number;
+    };
+    /** Recommended foreign accounts */
+    accounts?: string[];
+    /** Recommended foreign apps */
+    apps?: number[];
+    /** Recommended foreign assets */
+    assets?: number[];
+  };
+}
+```
+
+### Event Interface
+
+[ARC-28](./arc-0028.md) events are described using an extension of the original interface described in the ARC, with the addition of an optional struct field for arguments
+
+```ts
+interface Event {
+  /** The name of the event */
+  name: string;
+  /** Optional, user-friendly description for the event */
+  desc?: string;
+  /** The arguments of the event, in order */
+  args: Array<{
+    /** The type of the argument */
+    type: ABIType;
+    /** Optional, user-friendly name for the argument */
+    name?: string;
+    /** Optional, user-friendly description for the argument */
+    desc?: string;
+    /** If the type is a struct, the name of the struct */
+    struct?: StructName;
+  }>;
+}
+```
+
+### Type Interfaces
+
+The types defined in [ARC-4](./arc-0004.md) may not fully described the best way to use the ABI values as intended by the contract developers. These type interfaces are intended to supplement ABI types so clients can interact with the contract as intended.
+
+```ts
+/** An ABI-encoded type */
+type ABIType = string;
+
+/** The name of a defined struct */
+type StructName = string;
+
+/** Raw byteslice without the length prefixed that is specified in ARC-4 */
+type AVMBytes = "bytes";
+
+/** Mapping of named structs to the ABI type of their fields */
+interface StructFields {
+  [fieldName: string]: ABIType | StructFields;
+}
+```
+
+### Storage Interfaces
+
+These interfaces properly describe how app storage is access within the contract
+
+```ts
+/** Describes a single key in app storage */
+interface StorageKey {
+  /** Description of what this storage key holds */
+  desc?: string;
+  /** The type of the key */
+  keyType: ABIType | AVMBytes | StructName;
```

Comment:
Why does StorageKey have keyType? Isn't this for singular, known values?

Reply:
Yes but I figured it's still useful to know the type for things like explorers and frontends

Reply:
So it's intended to indicate the encoding of `key`? Maybe `keyEncoding` would be clearer? I don't think it benefits from consistency of naming with `StorageMap.keyType` since that seems of a different purpose - to indicate what sort of value to encode to combine with `StorageMap.key` in order to retrieve a value. Whereas `StorageKey.key` is already encoded and you may want to know how to interpret that value. So they're kind of opposites?

Reply:
Yeah, I'm a bit confused about what `keyType` means?

Reply:
Oh wait, I think I see.

So what should this and the `key` value be for a string value like we had in arc-32?

Needing to have key as a base64 reduces the human readability of these files... I wonder if keyType should allow for a raw string to be used?

Reply:
I think this would be the equivalent?

```typescript
          key: Buffer.from('my key', 'utf-8').toString("base64"),
          keyType: 'bytes',
```

Reply:
But then that doesn't let explorers etc. know that the key is actually a (non-ARC4) string?

Reply:
I think it would be confusing to have some keys be ASCII encoded and some keys be base64, thus everything base64 makes the most sense to me. I'm not sure if human readability is all that important. If you want to make it human readable just through it into Lora (once supported).

Reply:
Human readable is a nice feature though - makes it easy to see things at a glance, I've certainly found that useful in the past.

Regardless if that's supported or not, it probably makes sense to support a keyType of `utf-8`, which would either mean non-base64 or base64 `bytes`, but representing a UTF-8 string so the likes of Lora can show it properly.

Also, similarly we should probably have a key type of avm_address or similar to denote when you use the public key bytes of an address directly (i.e. non-ARC4 encoding).